Cannot get or set parameters with the type of PARAMETER_INTEGER_ARRAY

I have several parameters whose type are PARAMETER_INTEGER_ARRAY. It seems that rqt_reconfigure hasn't support this type yet, since I was trying to launch rqt_reconfigure, it failed with the following error messages:

> Traceback (most recent call last):
  File "/home/intel/workspace/ros2_other_ws/build/rqt_reconfigure/src/rqt_reconfigure/node_selector_widget.py", line 277, in _selection_changed_slot
    self._selection_selected(index_current, rosnode_name_selected)
  File "/home/intel/workspace/ros2_other_ws/build/rqt_reconfigure/src/rqt_reconfigure/node_selector_widget.py", line 228, in _selection_selected
    raise e
  File "/home/intel/workspace/ros2_other_ws/build/rqt_reconfigure/src/rqt_reconfigure/node_selector_widget.py", line 226, in _selection_selected
    item_widget = item_child.get_param_client_widget()
  File "/home/intel/workspace/ros2_other_ws/build/rqt_reconfigure/src/rqt_reconfigure/treenode_qstditem.py", line 94, in get_param_client_widget
    self._context, self._raw_param_name
  File "/home/intel/workspace/ros2_other_ws/build/rqt_reconfigure/src/rqt_reconfigure/param_client_widget.py", line 116, in __init__
    self._param_client.get_parameters(param_names),
  File "/home/intel/workspace/ros2_other_ws/build/rqt_reconfigure/src/rqt_reconfigure/param_api.py", line 89, in get_parameters
    for name, value in zip(names, get_params_response.values)
  File "/home/intel/workspace/ros2_other_ws/build/rqt_reconfigure/src/rqt_reconfigure/param_api.py", line 89, in <listcomp>
    for name, value in zip(names, get_params_response.values)
  File "/home/intel/workspace/ros2_ws/install/rclpy/lib/python3.6/site-packages/rclpy/parameter.py", line 121, in from_parameter_msg
    return cls(param_msg.name, type_, value)
  File "/home/intel/workspace/ros2_ws/install/rclpy/lib/python3.6/site-packages/rclpy/parameter.py", line 132, in __init__
    raise ValueError("Type '{}' and value '{}' do not agree".format(type_, value))
ValueError: Type 'Type.INTEGER_ARRAY' and value 'array('q', [640, 480])' do not agree
[INFO] [PluginContainerWidget]: Connection to node=/realsense failed:
Type 'Type.INTEGER_ARRAY' and value 'array('q', [640, 480])' do not agree
[ERROR] [rqt_reconfigure]: Connection to node=/realsense failed:
Type 'Type.INTEGER_ARRAY' and value 'array('q', [640, 480])' do not agree


On the other hand, I was trying to get/set parameters by ros2 param. Everything works well.

>$ros2 param set my_node my_param [640,480]
Set parameter successful
>$ros2 param get my_node my_param
Integer values are: array('q', [640, 480])

--- a/src/rqt_reconfigure/param_editors.py
+++ b/src/rqt_reconfigure/param_editors.py
@@ -34,6 +34,8 @@
 
 from decimal import Decimal
 
+import array
+import json
 import math
 import os
 
@@ -440,9 +442,84 @@ class DoubleEditor(EditorWidget):
         self.update(float('NaN'))
 
 
+class ArrayEditor(EditorWidget):
+    _update_signal = Signal(list)
+
+    def __init__(self, *args, **kwargs):
+        super(ArrayEditor, self).__init__(*args, **kwargs)
+        ui_str = os.path.join(
+            package_path, 'share', 'rqt_reconfigure', 'resource',
+            'editor_string.ui')
+        loadUi(ui_str, self)
+
+        if isinstance(self.parameter.value, array.array):
+            self._paramval_lineedit.setText(str(self.parameter.value.tolist()))
+        else:
+            self._paramval_lineedit.setText(str(self.parameter.value))
+
+        # Update param server when cursor leaves the text field
+        # or enter is pressed.
+        self._paramval_lineedit.editingFinished.connect(self.edit_finished)
+
+        # Make param server update text field
+        self._update_signal.connect(self._update_gui)
+
+        # Add special menu items
+        self.cmenu.addAction(self.tr('Set to Empty String')
+                             ).triggered.connect(self._set_to_empty)
+
+        if self.descriptor.read_only:
+            self._paramval_lineedit.setReadOnly(True)
+            self.cmenu.setEnabled(False)
+
+    def update_local(self, value):
+        super(ArrayEditor, self).update_local(value)
+        logging.debug('DoubleArrayEditor update_local={}'.format(value))
+        if isinstance(self.parameter.value, array.array):
+            self._update_signal.emit(value.tolist())
+        else:
+            self._update_signal.emit(value)
+
+    def _update_gui(self, value):
+        self._paramval_lineedit.setText(str(value))
+
+    def edit_finished(self):
+        logging.debug('ArrayEditor edit_finished val={}'.format(
+            self._paramval_lineedit.text()))
+        params_string = self._paramval_lineedit.text()
+        if self.parameter.from_parameter_msg:
+            params_string = params_string.replace('\'', '"')
+        params_list = json.loads(params_string)
+
+        if isinstance(self.parameter.value, array.array):
+            if self.parameter.value.typecode == 'q':
+                params_list = [int(val) for val in params_list]
+            else:
+                params_list = [float(val) for val in params_list]
+            self.update(array.array(self.parameter.value.typecode, params_list))
+        else:
+            if Parameter.Type.from_parameter_value(self.parameter.value) \
+                    == Parameter.Type.BOOL_ARRAY:
+                params_list = [bool(val) for val in params_list]
+            elif Parameter.Type.from_parameter_value(self.parameter.value) \
+                    == Parameter.Type.BYTE_ARRAY:
+                params_list = [bytes(val) for val in params_list]
+            else:
+                params_list = [str(val) for val in params_list]
+            self.update(params_list)
+
+    def _set_to_empty(self):
+        self.update('[]')
+
+
 EDITOR_TYPES = {
     Parameter.Type.BOOL: BooleanEditor,
     Parameter.Type.INTEGER: IntegerEditor,
     Parameter.Type.DOUBLE: DoubleEditor,
     Parameter.Type.STRING: StringEditor,
+    Parameter.Type.BOOL_ARRAY: ArrayEditor,
+    Parameter.Type.BYTE_ARRAY: ArrayEditor,
+    Parameter.Type.INTEGER_ARRAY: ArrayEditor,
+    Parameter.Type.DOUBLE_ARRAY: ArrayEditor,
+    Parameter.Type.STRING_ARRAY: ArrayEditor,
 }

--- a/src/rqt_reconfigure/param_editors.py
+++ b/src/rqt_reconfigure/param_editors.py
@@ -32,9 +32,8 @@
 #
 # Author: Isaac Saito, Ze'ev Klapow
 
-from decimal import Decimal
-
 import array
+from decimal import Decimal
 import json
 import math
 import os
@@ -488,7 +487,7 @@ class ArrayEditor(EditorWidget):
             self._paramval_lineedit.text()))
         params_string = self._paramval_lineedit.text()
         if self.parameter.from_parameter_msg:
-            params_string = params_string.replace('\'', '"')
+            params_string = params_string.replace("'", '"')
         params_list = json.loads(params_string)
 
         if isinstance(self.parameter.value, array.array):

